core/store/src/archive/cold_storage.rs

```diff
@@ -402,17 +403,28 @@ pub fn test_get_store_initial_writes(column: DBCol) -> u64 {
 fn get_keys_from_store(
     store: &Store,
     shard_layout: &ShardLayout,
+    tracked_shards: &Vec<ShardUId>,
     height_key: &[u8],
     block_hash_key: &[u8],
 ) -> io::Result<HashMap<DBKeyType, Vec<StoreKey>>> {
     let mut key_type_to_keys = HashMap::new();
+    let tracked_shard_ids: HashSet<ShardId> =
+        tracked_shards.iter().map(|shard_uid| shard_uid.shard_id()).collect();
 
     let block: Block = store.get_ser_or_err_for_cold(DBCol::Block, &block_hash_key)?;
+    let chunk_hashes = block
+        .chunks()
+        .iter_deprecated()
+        .map(|chunk_header| chunk_header.chunk_hash())
+        .collect::<Vec<ChunkHash>>();
     let chunks = block
         .chunks()
         .iter_deprecated()
-        .map(|chunk_header| {
-            store.get_ser_or_err_for_cold(DBCol::Chunks, chunk_header.chunk_hash().as_bytes())
+        .filter_map(|chunk_header| {
+            if !tracked_shard_ids.contains(&chunk_header.shard_id()) {
+                return None;
+            }
```

Comment:
The cold loop may be behind (e.g. by a few blocks because at best at final block or by many blocks due to compaction happening) . Is it certain that the tracked shards correspond to the shard layout that we're looking at here? In other words is it possible that we're using the currently tracked shards but trying to copy something from the old shard layout?

Reply:
Perhaps instead of this check you can just skip any chunks that are missing from the db? It seems more robust against the above and less robust against actual db errors being hidden and data loss.